Markup 2020 additional edits (#1858)
HTTPArchive/almanac.httparchive.org: src/content/en/2020/markup.md

How do we use HTML, then, how great of a foundation do we have? In the introductory section of the [2019 Markup chapter](../2019/markup#introduction), author [Brian Kardell](../2019/contributors#bkardell) suggested that for a long time, we haven't really known. There were some smaller samples. For example, there was [Ian Hickson's research](https://web.archive.org/web/20060203035414/http://code.google.com/webstats/index.html) (one of modern HTML's parents) among a few others, but until last year we lacked major insight into how we as developers, as authors, make use of HTML. In 2019 we had both [Catalin Rosu's work](https://www.advancedwebranking.com/html/) (one of this chapter's co-authors) as well as the 2019 edition of the Web Almanac to give us a better view again of HTML in practice.

> Last year's analysis was based on 5.8 million pages, of which 4.4 million were tested on desktop and 5.3 million on mobile. This year we analyzed 7.5 million pages, of which 5.6 million were tested on desktop and 6.3 million on mobile, using the [latest data](./methodology#websites) on the websites users are visiting in 2020. We do make some comparisons to last year but have also analyzed additional metrics for new insights. We've also tried to impart our own personalities and perspectives throughout the chapter.

Reviewer: Do you feel strongly about this one? I feel it somehow drags out the paragraph (for me the original reads faster), and I think we’d need to tweak further to avoid the double “also.”

Author: I could live with the original. I just don't see the connection between you trying to analysis more metrics and trying to impart your personalities. to be there's a stronger connection between the first two phrases and the third is separate. Unless you're trying to imply last year was boring cause they didn't impart enough of their personalities?

But, as I say, it's a small suggestion and I can live with the original.

```suggestion
Last year's analysis was based on 5.8 million pages, of which 4.4 million were tested on desktop and 5.3 million on mobile. This year we analyzed 7.5 million pages, of which 5.6 million were tested on desktop and 6.3 million on mobile, using the [latest data](./methodology#websites) on the websites users are visiting in 2020. We do make some comparisons to last year, but just as we've tried to analyze additional metrics for new insights, we've also tried to impart our own personalities and perspectives throughout the chapter.
```

      <tr>
        <th>Doctype</th>
        <th>Pages</th>

>         <th>Pages (%)</th>

Reviewer: I’ve commented on the general issue of changing the headers/labels in #467; this _seems_ fine, however.

Author: Yeah tried to standardise on this when I could since it was used in some places but not others.

<![endif]-->
```

> The above is a non-standard HTML conditional comment. While those have proven to be helpful in the past in order to tackle browser differences, they are historical for some time as Microsoft [dropped conditional comments](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/compatibility/hh801214(v=vs.85)) in Internet Explorer 10.

Reviewer: Just to flag this—I’m still not a native speaker but this reads odd. Is this really how one would say this? If you’re sure and this is a solid improvement, feel free to keep it—otherwise I suggest to leave this as is.

Author: I struggled with this one myself. Neither reads great IMHO. What about this?:

```suggestion
The above is a non-standard HTML conditional comment. While those have proven to be helpful in the past in order to tackle browser differences, they have been consigned to history for some time as Microsoft [dropped conditional comments](https://docs.microsoft.com/en-us/previous-versions/windows/internet-explorer/ie-developer/compatibility/hh801214(v=vs.85)) in Internet Explorer 10.
```

Reviewer: LGTM.

> Still, on the above percentile extremes, we found that web pages are using about 6 conditional comments in the 90th percentile, and 1 conditional comment while in the 10th percentile.  Most of the pages include them for helpers such as [html5shiv](https://github.com/aFarkas/html5shiv), [selectivizr](http://selectivizr.com/), and [respond.js](https://github.com/scottjehl/Respond). While being decentish and still active pages, our conclusion is that many of them were using obsolete CMS themes.

Reviewer: (Nice edit, thanks for adding all the detail.)

#### `details` and `summary`

> Another thing we were curious about was the use of the [`details` and `summary` elements](https://html.spec.whatwg.org/multipage/rendering.html#the-details-and-summary-elements), especially since 2020 brought [broad support](https://caniuse.com/details). Are they being used? Are they attractive for—even popular—among authors? As it turns out, only 0.39% of all tested pages are using them—although it's hard to gauge whether they were all used the correct way in exactly the situations when you need them.

Reviewer: No, that doesn’t work yet—with this edit, “although” doesn’t fit anymore. One could remove that last part (after the em dash), but then it takes away our interpretation. I suggest to keep this as it was.

A side note, “using these elements” could be a useful edit as “using them” starts to get unclear. If you could edit that then I wouldn’t need to track it (I had reviewed the chapter independently and am preparing another—small—PR).

Author: I have to read that first sentence several times to understand it - particularly with the last `"popular is the wrong word." bit.

What about dropping that but reverting the em-dash to a comma?

```suggestion
Another thing we were curious about was the use of the [`details` and `summary` elements](https://html.spec.whatwg.org/multipage/rendering.html#the-details-and-summary-elements), especially since 2020 brought [broad support](https://caniuse.com/details). Are they being used? Are they attractive for—even popular—among authors? As it turns out, only 0.39% of all tested pages are using them, although it's hard to gauge whether they were all used the correct way in exactly the situations when you need them.
```

Reviewer: Actually, I think the em dash works—in the old setting. What do you think of the old version but with an em dash instead of a comma? (In my view it helps.)

Author: ```suggestion
Another thing we were curious about was the use of the [`details` and `summary` elements](https://html.spec.whatwg.org/multipage/rendering.html#the-details-and-summary-elements), especially since 2020 brought [broad support](https://caniuse.com/details). Are they being used? Are they attractive for—even popular—among authors? As it turns out, only 0.39% of all tested pages are using them—although it's hard to gauge whether they were all used the correct way in exactly the situations when you need them—"popular" is the wrong word.
```

      </tr>
    </tbody>
  </table>
  <figcaption>{{ figure_link(caption="Adoption of the <code>details</code> and <code>summary</code> elements.", sheets_gid="1406534257", sql_file="pages_element_count_by_device.sql") }}</figcaption>
</figure>

### Probability of element use

> Taking another look at element popularity, how likely is it to find a certain element in the DOM of a page? Sure, `html`, `head`, `body` are present on every HTML page (even though [these tags are all optional](https://meiert.com/en/blog/optional-html/), browsers will automatically add them), making them common elements, but what other elements are to be commonly found?

Reviewer: (Works for me but now I wonder if the sentence gets a bit unwieldy. Even for my taste 😅)

Author: For me "surely" doesn't work here and "sure" is I think what you meant.

I also feel it's also important to realise we are analysing post-processed HTML (as we are using a JavaScript custom metic that works on the DOM, rather than looking at the HMTL returned). So even though they are optional, they will have been added for the purposes  of our analysis. I feel that's an important point to get across.

Reviewer 2: I agree. I discovered that html, head, body are there even if not in the raw html. So showing their percent is inaccurate. They are probably not 100% due to other errors.

Author: Could be some pages are txt or XML or other things other than HTML. So they won’t have html, head and body even post-processing.

Reviewer: I think the “sure” change is fine, but the additional sentence fragment makes the sentence quite long and, unwieldy. I think the old version works but if we want to add more to the sentence we may need to break it up. (I can make a suggestion later if need be.)

Author: I'm tired of arguing. Let's remove the offending clause

```suggestion
Taking another look at element popularity, how likely is it to find a certain element in the DOM of a page? Sure, `html`, `head`, `body` are present on every HTML page (even though [these tags are all optional](https://meiert.com/en/blog/optional-html/)), making them common elements, but what other elements are to be commonly found?
```

  <figcaption>{{ figure_link(caption="High probabilities of finding a given element in pages of the Web Almanac 2020 sample.", sheets_gid="184700688", sql_file="pages_element_count_by_device_and_element_type_present.sql") }}</figcaption>
</figure>

> Standard elements are those that are or were part of the HTML specification, but which ones are rare to find? In our sample, that would bring up the following:

Reviewer: I’m not sure this makes this clearer or gets it to read better. Do you feel strongly about this (why)? Suggest to leave this as it was.

Author: "Which ones are you really rarely to find" just does not read well as a native speaker. Maybe it's the alliteration of the r's.

```suggestion
Standard elements are those that are or were part of the HTML specification, but which ones are you rarely to find? In our sample, that would bring up the following:
```

Reviewer: How about keeping this but rephrasing the second sentence then? “Which ones are rare to find?” That may address both our points?

Author: ```suggestion
Standard elements are those that are or were part of the HTML specification. Which ones are rare to find? In our sample, that would bring up the following:
```

</figure>

We're including these elements to give an idea what elements may have gone out of favor. But while `dir` and `basefont` were last specified in XHTML 1.0 (2000) and are no longer part of HTML, the rare use of `rp` (which was mentioned [as early as 1998](https://www.w3.org/TR/1998/WD-ruby-19981221/#a2-4) and is [still part of HTML](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-rp-element)), may just suggest that Ruby markup is not very popular.

### Custom elements

The 2019 edition of the Web Almanac handled [custom elements](../2019/markup#custom-elements) by discussing several non-standard elements. This year, we found it valuable to have a closer look at custom elements. How did we determine these? Roughly by looking at [their definition](https://html.spec.whatwg.org/multipage/custom-elements.html#custom-elements-core-concepts), notably their use of a hyphen. Let's focus on the top elements, in this case elements used on ≥1% of all URLs in the sample:

<figure>
  <table>
    <thead>
      <tr>
        <th>Element</th>

>         <th>Pages</th>

Reviewer: Off the top of my head, I don’t think this is what the data are about. The occurrences here are *not* pages—they are the number of elements.

We need to be really careful with these labels—a small change in wording can change the whole meaning, as in this case. I strongly advise against just changing them in passing, and I’m also nervous that we (the authors and analysts) get pulled in again for which is going to be a lot more work than it looks like, for given the nature of these edits, we need to review the data sets and should really double-check what is being changed.

So while I get the intent here to make things more clear, ironically it comes with a significant risk of actually making things less clear, or even wrong. I wouldn’t touch the labels anymore.

Author: That's the biggest thing I think we need to clarify here! The chapter uses "Occurrences" in one place to mean pages (i.e. multiple occurrences on a page still only count as 1), and in other  places to actually mean "Occurrences" (i.e. multiple occurrences on a page count each time).

Looking at both the data and the query I've attempted to clarify which really are occurrences and which are pages. It's entirely possible I got it wrong, so yes it would be good for @Tiggerito to review, but there is little doubt in my mind that the current chapter is presenting incorrect data.

For example this one has the following in the chapter:

![image](https://user-images.githubusercontent.com/10931297/103458049-4cddb580-4cfc-11eb-91b1-d3bbadceeee5.png)

Which relates to this data [from Sheets](https://docs.google.com/spreadsheets/d/1Ta7amoUeaL4pILhWzH-SCzMX9PsZeb1x_mwrX2C4eY8/edit#gid=184700688) 

![image](https://user-images.githubusercontent.com/10931297/103458064-726abf00-4cfc-11eb-9da1-ca89612bb6d8.png)

Therefore it seems clear that "occurrences" here means pages? But that is not clear to the reader.
As a third check[ the query](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2020/03_Markup/pages_element_count_by_device_and_custom_dash_elements.sql) also says this is pages.

Later on you have this:

![image](https://user-images.githubusercontent.com/10931297/103458092-afcf4c80-4cfc-11eb-9ee9-d52f68a314ff.png)

Now we don't have 10million pages so we know "Occurrences" here does not mean pages.

I feel this double usage of the same word to mean two things is NOT clear and you are presenting incorrect data to the readers. And the fact you and I aren't even sure which it is nails that point for me!

So I feel this is something we MUST correct. As I say I've spent considerable time looking at the data and the queries and _think_ I've got them right, but could well be wrong!

Reviewer 2: The first one is from the script:

pages_element_count_by_device_and_element_type_present.sql

The word present is meant to indicate I'm just counting if there are any. Percent of pages they are used on.

The second one comes from:

pages_element_count_by_device_and_element_type_frequency.sql

Where the word frequency is meant to indicate a full count. So that would be occurrences.

Author: OK so just to be 100% clear, you agree changing this line (which is about the first example) from “Occurrences” to “Pages”, like I’ve done in this PR, is correct?

The second one (headings table) is further down the page and I just included in this comment to make my point that we’re currently using the same word to mean different things - which is the primary thing I’m trying to correct in this PR (Qing with a few other smaller things). I agree this second “headings” table should stay as “Occurrences” as clearly it’s not pages since there more than the 7.5 million pages we check across mobile and desktop 🙂

Reviewer 2: Correct. I think the first example should say Pages as you did.

Reviewer: Hey—first, I think this is _awesome_ to look into improving the chapter. What I’d like to raise is, is what we’re improving here worth the effort (involving time of currently 3 people) and the risk (of something slipping through)?

In my mind this is a clear “no”—we have no indication that something is wrong, AFAIK no one complained about the wording, and generally it seems this is nice, yes, but also rather minor.

I respect other views here and I’m also willing to support this, but I am doubtful and concerned.

A side note, I want to underline the suggestion to give people a heads-up where possible, and I also suggest to break up PRs where possible. The header/label topic would feel much better to me had it been separated from the other changes—now those changes will need to wait, and so does another PR I was prepared to get out (I might get it ready regardless now, accepting a light risk of conflicts as this also needs to get out).

Author: As I stated when opening this PR, the motivation for this was not to endlessly tweak or criticise this chapter. During the first translation it was noticed we still had open TODOs in this chapter. This left me with one of four choices"
1. Ignore them and leave open TODOs in each translation - as they are in the Dutch version. If we ever addressed these in the future then we'd have to revisit every translation. I don't think this was a great option to be honest.
2. Assume all the TODOs were in fact done and close them. The chapter is what it is, and we were obviously happy to publish so the TODOs were not that important. That was an option but without looking at the remaining TODOs - which were obviously not addressed sufficiently to warrant removing them.
3. Ask Authors to investigate them. As far as I'm concerned this had been asked once before.
4. Spend considerable time looking through each TODO, closing out the ones that either had been addressed or were not a big deal (e.g. the document size one), and make suggestions to fix the other ones so they could be closed out. This is the choice I made despite it being the most effort on my part. Perhaps in hindsight this was the wrong choice.

> In my mind this is a clear “no”—we have no indication that something is wrong

We have a clear indication something is wrong! Occurrences in one place means number of elements, and in other means number of pages. You then argued that "occurrences here are not pages—they are the number of elements." but both my and @Tiggerito 's analysis shows that is incorrect — it means pages. I'd prefer each chapter to not knowingly continue to publish incorrect information and have put considerable the effort in to correct it, rather than just raising an issue and asking others to do the work.

> A side note, I want to underline the suggestion to give people a heads-up where possible, and I also suggest to break up PRs where possible. The header/label topic would feel much better to me had it been separated from the other changes—now those changes will need to wait, and so does another PR I was prepared to get out (I might get it ready regardless now, accepting a light risk of conflicts as this also needs to get out).

Yes perhaps this was a mistake of mine. I did break up the table reformats (#1587), as didn't feel those required approval, and to keep this PR small. I thought the wording changes were minor enough to include in this and have them reviewed at same time rather than make you review two PRs. I guess I was wrong about this.

I also like the improvements in your PR, despite there being no heads up and no complaints making them necessary, so happy to take them. I wish we'd closed out the one remaining issue here and merged this before you'd opened that though.

  <figcaption>{{ figure_link(caption="The 14 most popular custom elements.", sheets_gid="770933671", sql_file="pages_element_count_by_device_and_custom_dash_elements.sql") }}</figcaption>
</figure>

These elements come from three sources: [Yandex Metrica](https://metrica.yandex.com/about) (`ym-`), an analytics solution we also saw last year; [Slider Revolution](https://www.sliderrevolution.com/) (`rs-`), a WordPress slider, for which there are more elements to be found near the top of the sample; and [Wix](https://www.wix.com/) (`wix-`), a website builder.

Other groups that stand out include [AMP markup](https://amp.dev/) with `amp-` elements like `amp-img` (11,700 pages), `amp-analytics` (10,256) and `amp-auto-ads` (7,621), as well as [Angular](https://angular.io/) `app-` elements like `app-root` (16,314), `app-footer` (6,745), and `app-header` (5,274).

### Obsolete elements

> There are more questions to ask about the use of HTML including the use of obsolete elements, which are elements like `applet`, `bgsound`, `blink`, `center`, `font`, `frame`, `isindex`, `marquee`, or `spacer`.

Reviewer: I suggest to add a comma after “HTML” now, and to consider putting the last sentence fragments in parentheses, like “…obsolete elements (elements like …)”.

Author: SGTM!

```suggestion
There are more questions to ask about the use of HTML, including the use of obsolete elements (which are elements like `applet`, `bgsound`, `blink`, `center`, `font`, `frame`, `isindex`, `marquee`, or `spacer`).
```

    <thead>
      <tr>
        <th>Element</th>

>         <th>Pages</th>

Reviewer: See above (#467)—here, too, this would be wrong. This information does *not* relate to pages.

We should fix “Occurances” though—strangely, however, I can’t spot this misspelled variant in the live version of the chapter.

Author: @Tiggerito can you confirm?

Looking at [this part of the query](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2020/03_Markup/pages_element_count_by_device_and_obsolete_elements.sql):

```sql
  COUNT(0) AS freq,
  AS_PERCENT(COUNT(0), total) AS pct_pages_with_obsolete_elements,
  AS_PERCENT(COUNT(0), SUM(COUNT(0)) OVER (PARTITION BY _TABLE_SUFFIX)) AS ratio_compared_to_all_obsolete_elements
```

If we're dividing by total (which is based on pages) as in the second line then surely `COUNT(0)` must be pages too?

Author: Oh and btw looks like I introduced the misspelling of Occurances in #1857 when converting from Markdown tables to HTML (to allow correct alignment of numeric values). I deliberately did that separately to this PR to keep this PR on the edits.

Reviewer 2: Yes, I believe this one is pages based. It's only looking to see if an obsolete element exists on a page, nut how many times it is there.

### Top attributes

> Similar to the section on the most [popular elements](#top-elements), this section delves into the most popular attributes on the web. Given how important the `href` attribute is for the web itself, or the `alt` attribute in order to make information [accessible](./#images-and-their-text-alternatives), would these be most popular attributes?

Reviewer: I think this was meant to go to the Accessibility chapter—can you update?

Author: Oops! Good spot!

```suggestion
Similar to the section on the most [popular elements](#top-elements), this section delves into the most popular attributes on the web. Given how important the `href` attribute is for the web itself, or the `alt` attribute in order to make information [accessible](./accessibility#images-and-their-text-alternatives), would these be most popular attributes?
```

    <thead>
      <tr>
        <th>Content attribute value</th>

>         <th>Pages</th>

Reviewer: Same general problem as mentioned before and no, “pages” would be inaccurate.

Author: Actually this one I may be wrong. But then I feel the Pages(%) heading is inaccurate?

![image](https://user-images.githubusercontent.com/10931297/103458328-d3939200-4cfe-11eb-9779-8f6306dac694.png)

Though to be honest for this one, as there's usually only one per page it's probably the same for both!

@Tiggerito can you confirm?

Reviewer 2: One per page and the report includes if it is missing (blank). So I say it's pages and % pages.

    <thead>
      <tr>
        <th>Favicon format</th>

>         <th>Pages</th>

Reviewer: Same general problem as mentioned before and no, “pages” would be inaccurate.

Author: Think you might be right but then Pages (%) is also inaccurate?

![image](https://user-images.githubusercontent.com/10931297/103458355-0473c700-4cff-11eb-9cbe-d4332013215b.png)

Reviewer 2: Same here, one result per page, so pages and % pages is accurate. I think in this case it will pick the first matching tag on the page.

  sql_file="pages_markup_by_device.sql"
) }}

<figure>
  <table>
    <thead>
      <tr>
        <th>Button types</th>

>         <th>Pages</th>

Reviewer: Same general problem as mentioned before and no, “pages” would be inaccurate.

Author: Ah this one I'm definitely wrong! 

```suggestion
        <th>Occurrences</th>
```
